### PR TITLE
ED-159: Ipn system

### PR DESCRIPTION
**Pull request by BDavid57**
### Overview

This pull request introduces an IPN system

### Details

- Updated application event-type.enum
- Created an IPN module
- Created the class PaymentAgreementsIpnNotificationsHandler
- Updated the SignPaymentAgreementEventTopics class in order to accept metadata and also refactored it to decode the data using a multiversx library instead of decoding it manually
- Updated the sign agreement event handler to save the metadata in the database
- Refactored agreement trigger event handler

### CODE REVIEW

File: src/features/payment-agreements/payment-agreement-members.service.ts [hunk 1]
```diff
@@ -20,48 +20,48 @@ export class PaymentAgreementMembersService {
...
-    .find({ member: address })
+      .find({ member: address });
     
-    const aggrementIds = memberships.map(item => item.internalAgreementId);
+    const agreementIds = memberships.map(item => item.internalAgreementId);
 
-    const operationsCount = await this.repository.model.find({ _id: { $in: aggrementIds } }).countDocuments({})
-    const itemsPerPage = PaymentAgreementMembersService.ITEMS_PER_PAGE
-    const numberOfPages = Math.ceil(operationsCount / itemsPerPage)
+    const operationsCount = await this.repository.model.find({ _id: { $in: agreementIds } }).countDocuments({});
```
Comment (andreigorgan) [inline src/features/payment-agreements/payment-agreement-members.service.ts:27]: Should be in `agreementsRepository` repository, a method called `getAgreementsByIdsCount`
. Note that you are not inquiring the wrong repository here.

File: src/features/payment-agreements/payment-agreement-members.service.ts [hunk 2]
```diff
@@ -20,48 +20,48 @@ export class PaymentAgreementMembersService {
...
+    const itemsPerPage = PaymentAgreementMembersService.ITEMS_PER_PAGE;
+    const numberOfPages = Math.ceil(operationsCount / itemsPerPage);
 
     const allSignedAgreements: PaymentAgreement[] = await this.agreementsRepository.model
```
Comment (andreigorgan) [inline src/features/payment-agreements/payment-agreement-members.service.ts:31]: This should be in a method in `agreementsRepository`,  e.g.`findAgreementsByIdsPerPage`

File: src/features/payment-agreements/payment-agreement-members.service.ts [hunk 3]
```diff
@@ -20,48 +20,48 @@ export class PaymentAgreementMembersService {
...
       numberOfPages: numberOfPages
-    }
+    };
   }
 
   async findAgreementMembers(id: Types.ObjectId, pagination: PaginationParams = new PaginationParams()) {
-    const operationsCount = await this.repository.model.find({ internalAgreementId: id }).countDocuments({})
-    const itemsPerPage = PaymentAgreementMembersService.ITEMS_PER_PAGE
-    const numberOfPages = Math.ceil(operationsCount / itemsPerPage)
+    const operationsCount = await this.repository.model.find({ internalAgreementId: id }).countDocuments({});
```
Comment (andreigorgan) [inline src/features/payment-agreements/payment-agreement-members.service.ts:47]: This should be a repository method, see above comments.

File: src/features/payment-agreements/payment-agreement-members.service.ts [hunk 4]
```diff
@@ -20,48 +20,48 @@ export class PaymentAgreementMembersService {
...
+    const itemsPerPage = PaymentAgreementMembersService.ITEMS_PER_PAGE;
+    const numberOfPages = Math.ceil(operationsCount / itemsPerPage);
     
     const allMemberships = await this.repository.model
```
Comment (andreigorgan) [inline src/features/payment-agreements/payment-agreement-members.service.ts:51]: This should be in a repository method, see above comments.

File: src/features/payment-agreements/payment-agreement-members.service.ts [hunk 5]
```diff
@@ -20,48 +20,48 @@ export class PaymentAgreementMembersService {
...
-    .find({ internalAgreementId: id })
-    .skip(pagination.skip)
-    .limit(pagination.limit)
-    .sort({ createdAt: 'desc' })
+      .find({ internalAgreementId: id })
+      .skip(pagination.skip)
+      .limit(pagination.limit)
+      .sort({ createdAt: 'desc' });
 
     return {
```
Comment (andreigorgan) [inline src/features/payment-agreements/payment-agreement-members.service.ts:57]: We should define a common pagination model and use that in all paginated items. E.g.
```
type PaginatedResponse<T> {
    items: T[];
   totalPages: int;
   currentPage: int;
... etc
}
```

The structure doesn't have to be exactly this but you get the point

File: src/features/payment-agreements/payment-agreement-members.service.ts [hunk 6]
```diff
@@ -20,48 +20,48 @@ export class PaymentAgreementMembersService {
...
       memberships: allMemberships,
       numberOfPages: numberOfPages
-    }
+    };
   }
 
   async updateLastChargedAt(member: string, date: Date){
-    const newCharge = { $set: { lastSuccessfulCharge: date }}
+    const newCharge = { $set: { lastSuccessfulCharge: date } };
```
Comment (andreigorgan) [inline src/features/payment-agreements/payment-agreement-members.service.ts:64]: This should be part of the repository, The repository should know how to construct the update request and which update method to use.

**Review by andreigorgan — APPROVED** [review]

File: src/features/payment-agreements/payment-agreement-members.service.ts (hunk 5 above)
Comment (BDavid57) [reply src/features/payment-agreements/payment-agreement-members.service.ts:56]: This is done on another branch

**Pull request merged**